### ISSUE TITLE
Plymouth does not show password prompt with encrypted root

### ISSUE THREAD

**Issue by Slabity** [NixOS/nixpkgs#26722]
## Issue description
Plymouth has the ability to ask the user for their password to decrypt their root device if necessary. Currently, Plymouth runs after the user already decrypts their device in the TTY, and as far as I can tell there is no option to override this.

### Steps to reproduce
1. Encrypt your root with LUKS.

2. Enable plymouth:

```boot.plymouth.enable = true```

3. Reboot. The init process will request your password before plymouth runs.

## Technical details

* System: NixOS 17.09pre108553.0011f9065a (HummingBird)
* Nix version: 1.11.9
* Nixpkgs version: 17.09pre108553.0011f9065a
* Sandboxing enabled: false

**Comment by abbradar** [comment]
I remember myself trying to implement this some time ago; this would probably need us moving from shell scripts based initrd to systemd. Maybe I'll take another look at this...

**Comment by gilligan** [comment]
Has anything ever happened in this regard? Does anyone know?

**Comment by Slabity** [comment]
Unfortunately I was unable to get this working. I ended up disabling plymouth entirely. Hopefully someone else has a better idea on how to get this working.

**Comment by jtojnar** [comment]
Nope @hedning, this issue has been there for a long time.

**Comment by nioncode** [comment]
Ideally, we should have flicker free boot like Fedora has now (see https://hansdegoede.livejournal.com/19224.html and https://hansdegoede.livejournal.com/19673.html).

**Comment by andersk** [comment]
Still an issue. I suspect this will require #72401 (systemd in initramfs) and #99011 (nixos/plymouth: Add label plugin and a font to the initrd).

**Comment by cole-h** [comment]
May also be influenced by https://github.com/NixOS/nixpkgs/pull/88789.

**Comment by TLATER** [comment]
#99011 was merged, for reference.

**Comment by anthr76** [comment]
Since the above was implemented is anything else blocking this from being resolved?

**Comment by WillPower3309** [comment]
> Since the above was implemented is anything else blocking this from being resolved?

Afaik #170136 will fix this issue

**Comment by WillPower3309** [comment]
Anyone test with latest nixpkgs master? Afaik it should work based on the systemd in stage 1 milestone progress

**Comment by iclanzan** [comment]
I am on commit https://github.com/NixOS/nixpkgs/commit/934e076a441e318897aa17540f6cf7caadc69028, added `boot.plymouth.enable = true` and Plymouth didn’t even come up at all. Do I need other settings now? Should I try an even fresher commit?

**Comment by iclanzan** [comment]
I am seeing these lines in the journal that mention Plymouth:
```
May 10 19:03:41 ken systemd[1]: Forward Password Requests to Plymouth was skipped because of a failed condition check (ConditionPathExists=/run/plymouth/pid).
May 10 19:03:42 ken systemd[1]: Starting Tell Plymouth To Write Out Runtime Data...
May 10 19:03:42 ken systemd[1]: Finished Tell Plymouth To Write Out Runtime Data.
```

The "condition check" message is repeated half a dozen times.

**Comment by WillPower3309** [comment]
> I am seeing these lines in the journal that mention Plymouth:
> ```
> May 10 19:03:41 ken systemd[1]: Forward Password Requests to Plymouth was skipped because of a failed condition check (ConditionPathExists=/run/plymouth/pid).
> May 10 19:03:42 ken systemd[1]: Starting Tell Plymouth To Write Out Runtime Data...
> May 10 19:03:42 ken systemd[1]: Finished Tell Plymouth To Write Out Runtime Data.
> ```
> 
> The "condition check" message is repeated half a dozen times.

@dasJ does this have implications wrt #170136 ?

**Comment by nomeata** [comment]
Just gave it a shot, release-22.05, with
```
boot.initrd.systemd.enable = true;
boot.plymouth.enable = true;
```
There was systemd in stage1, and there was a job that said something about starting plymouth before it would ask for the password, but plymouth did not actually come up, and after a short pause the screen was reset and more textual start up message appeared, including a textual password prompt. (This was my first experiment with systemd in stage1, and other things were strange, so I reverted for now.)

**Comment by naufik** [comment]
On 22.05, I managed to have a minimal nixos-branded plymouth password prompt using the following config:

```nix
boot.plymouth = {
  enable = true;
  theme = "breeze";
}

boot.initrd.systemd.enable = true;
```

The text "Starting Plymouth" appears beforehand, and afterwards a graphical prompt appears.

According to the [documentation](https://search.nixos.org/options?channel=22.05&show=boot.plymouth.themePackages&from=0&size=50&sort=relevance&type=packages&query=boot.plymouth) for `boot.plymouth.themePackages` a package is only specified by default for theme "breeze".

I got the same result from testing the code [above](https://github.com/NixOS/nixpkgs/issues/26722#issuecomment-1147735675), getting a text prompt after "Starting Plymouth". So potentially it's the default theme `"bgrt"` not including a graphical prompt.

**Comment by TLATER** [comment]
Can this issue be closed? Or should we set `breeze` by default, or update the `plymouth.enable` documentation to point this quirk out?

**Comment by zeorin** [comment]
Not working for me with @nomeata's settings… Nixos 22.05.

Plymouth starts, (though it's very slow and the wrong resolution), and shows me the passphrase prompt. When I enter the passphrase the prompt disappears and nothing happens.

When I press Escape the console output says that we're waiting for a job for the a disk.

I have a somewhat complicated setup, I have 2 disks that I wanted to unlock with one passphrase, which I wanted to enter only once, so I have 3 encrypted partitions.

The first is unlocked with a passphrase, and it is the keyfile for the following 2 encrypted partitions. The first of these is the root filesystem. The second is just a data drive:

```nix
{ config, lib, pkgs, ... }:

{
  ...

  boot = {
    ...

    initrd.luks.devices = {
      cryptkey = {
        device = "/dev/disk/by-uuid/6b17a213-6987-4a8e-b609-5243f6ba1467";
        preLVM = true;
      };
      cryptroot = {
        device = "/dev/disk/by-uuid/556cb835-419a-48b6-a081-36d2998d9c57";
        keyFile = "/dev/mapper/cryptkey";
        preLVM = true;
        allowDiscards = true;
      };
    };

    initrd.systemd.enable = true;

    plymouth = {
      enable = true;
      theme = "breeze";
    };

  };

  ...

  fileSystems."/data" = {
    device = "/dev/disk/by-uuid/6ee6e25c-fe6f-4c50-b7fb-985260cf8ca9";
    encrypted = {
      enable = true;
      label = "cryptdata";
      blkDev = "/dev/disk/by-uuid/14924ada-f427-411b-b426-e9db44ab0752";
      keyFile = "/dev/mapper/cryptkey";
    };
  };

  ...
}

```

**Comment by TLATER** [comment]
@zeorin that sounds more like a bug with the implementation that doesn't permit your setup or at least is backwards-incompatible with it, and not a flat out missing implementation like it used to be. Perhaps that should be handled in a aeparate issue?

**Comment by GoldsteinE** [comment]
It also doesn’t work when `initrd.luks.yubikeySupport = true;`, but that’s probably an another issue.

**Comment by Majiir** [comment]
I have an encrypted ZFS root. With this config

```nix
boot.plymouth.enable = true;
boot.initrd.systemd.enable = true;
```

I get a textual prompt for the password. But if I include the changes from https://github.com/NixOS/nixpkgs/pull/179619, I get a graphical prompt, even with the default `bgrt` theme.

It sure does take a long time for udev initialization to finish and for Plymouth to take over, though.

**Comment by nomeata** [comment]
A new release, a new try, with some progress! With 22.11 and
```
boot.initrd.systemd.enable = true;
boot.plymouth.enable = true;
```
after grub I first get a few lines of text, including something akin to “Starting plymouth”, which takes a few moments. Then I do get a graphical password prompt from plymouth (success!), and it unlocks my disk. Text mode briefly pops up again, before X comes up. I then have no Wifi for some reason that I didn’t debug further (probably due to systemd stage 1?).

Even if I could fix the wifi issue, it’s still noticable slower than the text mode passphrase prompt, so in that form not yet worth it. I guess I am spoiled and hope for a fast and flickerless experience from grub all the way through to X, which is what I had in Debian before switching to Nix OS :-)

**Comment by Zahrun** [comment]
I configured NixOS 22.11 with
```
boot.initrd.systemd.enable = true;
boot.plymouth.enable = true;
boot.plymouth.theme = "breeze";
environment.systemPackages = with pkgs; [
  plymouth
  breeze-plymouth
];
```
After nixos-rebuild switch and reboot, I am still welcomed with the GRUB "Enter passphrase for hd0,msdos1" text password input. What am I doing wrong?

**Comment by TLATER** [comment]
If you're using grub's cryptomount to unlock your luks you can't use plymouth for password prompt, because plymouth is part of the initrd and requires the kernel to already be running.

It's arguably better to do what you're doing though, since it means you can encrypt your `/boot` as well, albeit at a performance cost/limitation to LUKS1, or having to type passwords multiple times. It also only really helps if you use secure boot to verify grub.

People here let the initrd unlock their devices, which means that the kernel must have already started. It's the setup most commonly used by Linux distros out there, but requires leaving `/boot` unencrypted and because the initrd is never verified with secure boot it leaves your system vulnerable to evil maid attacks that swap in a malicious initrd.

Poettering has recently started a crusade against this poor state of affairs if you're curious about a potential better future approach: https://0pointer.de/blog/brave-new-trusted-boot-world.html

Probably will take a long time before adoption in NixOS though.

**Comment by Zahrun** [comment]
That you @TLATER, I was able to reproduce the setup with /boot unencrypted and / encrypted, at first I was surprised that NixOS was not even asking for a password to boot. For that I had to comment the line
`boot.initrd.luks.devices."luks-xxx".keyfile = "/crypto_keyfile.bin"`.

I was always confused about why different distros had differences in luks unlocking, so thank you helping me understand the difference between grub unlocking and initrd unlocking.

However, even with 
```
boot.initrd.systemd.enable = true;
boot.plymouth.enable = true;
boot.plymouth.theme = "breeze";
environment.systemPackages = with pkgs; [
  plymouth
  breeze-plymouth
];
```
I get a line `Show Plymouth Boot Screen`, and then the text prompt as below:
![VirtualBox_NixOS plymouth_24_01_2023_01_17_55](https://user-images.githubusercontent.com/10415894/214135558-e75b118f-d0dc-4bd4-8335-206564284fc6.png)

I get the same result if I use bgrt theme for plymouth. What am I doing wrong this time?

**Comment by TLATER** [comment]
I'd recommend asking for support on discourse rather than github tickets, but the related ticket #32556 may have some hints.

I've definitely seen it work with that configuration before, though as of 21.11 setting the theme to breeze doesn't seem necessary anymore, I guess the default theme has gained password prompt support.

**Comment by Majiir** [comment]
I have Plymouth showing the password prompt for both ZFS and LUKS encrypted volumes on 23.05. The Discourse thread also indicates the issue is resolved.

Is this two-line config snippet sufficient to close this issue? Or is it a problem that `boot.initrd.systemd` is required?

```nix
boot.initrd.systemd.enable = true;
boot.plymouth.enable = true;
```

**Comment by nomeata** [comment]
I can (still) confirm that it works, and that this issue can likely be closed.

I’m not using it, though, it increases my laptop boot time (from grub to X) from 9s to 17s. Yes, these numbers are amazing to what we had around 2000, but still :-)

Also there is unfortunately still flicker, and the status messages from systemd are printed in normal text mode. But flicker-free boot is a topic for a different issue… :-)

**Comment by TLATER** [comment]
> after adding `boot.kernelParams = [ "quiet" "udev.log_level=3" ];` there is no more messages in text mode

Is this actually what other distros do? I feel like it's an anti-pattern because it breaks plymouth's escape-to-see-logs feature.

**Comment by Majiir** [comment]
Closing this as completed. For future readers, LUKS+Plymouth works if you set `boot.initrd.systemd.enable = true`. This enables the systemd initrd, which replaces the older script-based stage 1 initrd.

See https://github.com/NixOS/nixpkgs/issues/44965 for a similar issue about ZFS+Plymouth.

**Comment by eclairevoyant** [comment]
> > after adding `boot.kernelParams = [ "quiet" "udev.log_level=3" ];` there is no more messages in text mode
> 
> Is this actually what other distros do?

[Yes.](https://gitlab.archlinux.org/archlinux/packaging/packages/grub/-/blob/main/grub.default?ref_type=heads#L6)

**Comment by ElvishJerricco** [comment]
@Majiir I disagree with closing this. Systemd initrd is still marked experimental and is not feature complete. The default initrd in NixOS has this issue.

**Comment by WillPower3309** [comment]
For those running a system with a root tmpfs, you'll need to persist the `/run/plymouth` directory

**Comment by ElvishJerricco** [comment]
@WillPower3309 that doesn't make much sense, because `/run` is a tmpfs anyway and is not persisted on *any* system.

**Comment by WillPower3309** [comment]
> @WillPower3309 that doesn't make much sense, because `/run` is a tmpfs anyway and is not persisted on *any* system.

I see you're right, and either way Plymouth starts before the mount, but why would Plymouth output:
```
Forward Password Requests to Plymouth was skipped because of a failed condition check (ConditionPathExists=/run/plymouth/pid).
```

**Comment by ElvishJerricco** [comment]
@WillPower3309 I'm not exactly sure why that happens, but it's a red herring. I see that on my system (during stage 2) and plymouth's password prompt works fine (even for passwords in stage 2)

**Comment by ElvishJerricco** [comment]
I wonder if this is a `simpledrm` problem. Plymouth refuses to use `simpledrm` devices for the first 8 seconds during boot, which can seem like plymouth isn't working. You can use `plymouth.use-simpledrm` in kernel params to force it to use that.

**Comment by musjj** [comment]
I'm also occasionally experiencing this even with `boot.initrd.systemd` enabled. Like mentioned above, it looks like that there's some kind of race condition. 

Sometimes the Plymouth password prompt is displayed just fine, sometimes it doesn't.

Does anyone know any workarounds?

**Comment by eclairevoyant** [comment]
[This thread](https://discourse.nixos.org/t/i-found-a-fix-for-plymouth-not-displaying-but-im-not-sure-who-to-tell/77247?u=waffle8946) proposes a fix.
